applicationset/services/github_metrics.go

```diff
@@ -0,0 +1,257 @@
+package services
+
+import (
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	log "github.com/sirupsen/logrus"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+)
+
+// Doc for the GitHub API rate limit headers:
+// https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28#checking-the-status-of-your-rate-limit
+
+// Metric names as constants
+const (
+	githubAPIRequestTotalPerAppSetMetricName       = "argocd_github_api_requests_total_per_appset"
+	githubAPIRequestDurationPerAppSetMetricName    = "argocd_github_api_request_duration_seconds_per_appset"
+	githubAPIRateLimitRemainingPerAppSetMetricName = "argocd_github_api_rate_limit_remaining_per_appset"
+	githubAPIRateLimitLimitPerAppSetMetricName     = "argocd_github_api_rate_limit_limit_per_appset"
+	githubAPIRateLimitResetPerAppSetMetricName     = "argocd_github_api_rate_limit_reset_per_appset"
+	githubAPIRateLimitUsedPerAppSetMetricName      = "argocd_github_api_rate_limit_used_per_appset"
+	githubAPIRateLimitResourcePerAppSetMetricName  = "argocd_github_api_rate_limit_resource_per_appset"
```

Comment:
Can we drop the `_per_appset` suffix and instead just document that the `appset_name` and `appset_namespace` labels are available?

Reply:
+1